### PR TITLE
CMR Migrations - consuming.

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

Note: this PR is the first of many to introduce CMR migrations into
juju, this is the first one that introduces importation of remote
applications. There are no QA steps for this as nothing is done with
the output of the remoteapplications yet.

----

Included in the commit is the addition of the cmr entities for
consuming a remote application. The code is hidden behind a feature
flag to ensure that we don't break any imports that already exist.

Much more work is required to ensure that we also import all the
other entities required.

Part of the change is also the introduction of go mocks in the state
package. This allows us to reach a higher level of coverage without
the need of the persistence layer.

Not only are the tests in mocks, but additionally added to the existing
integration tests that we have in state.

## QA steps

Test pass, offering part will have full QA steps.

### CODE REVIEW

File: state/migration_export.go [hunk 1]
```diff
@@ -1121,6 +1121,73 @@ func (e *exporter) relations() error {
...
+	AllRelationNetworks() ([]RelationNetworks, error)
+}
+
+// RelationNetworksModel defines an inplace usage for adding a relation networks
+// to a model.
+type RelationNetworksModel interface {
+	AddRelationNetwork(description.RelationNetworkArgs) description.RelationNetwork
+}
+
+func exportRelationNetworks(state RelationNetworksSource, model RelationNetworksModel) error {
```
Comment (manadart) [inline state/migration_export.go]: For posterity... as we discussed:
https://play.golang.org/p/GC1tbWd272e

File: Gopkg.toml [hunk 2]
```diff
@@ -62,8 +62,9 @@
   name = "github.com/juju/bundlechanges"
 
 [[constraint]]
-  revision = "7eee51ddb1f5033244c6501ec83b399af9d23861"
+  revision = "be85422a108107ae554950534717443e97dc70d0"
   name = "github.com/juju/description"
+  source = "github.com/SimonRichardson/description"
```
Comment (SimonRichardson) [inline Gopkg.toml]: Once this PR has been approved from a code point of view and it handles the descriptions correctly, I'll merge the description package first, before merging this PR.

The should ensure that the churn on description is limited.

**Comment by SimonRichardson** [comment]
https://trello.com/c/yWdmtjaP/1811-cmr-migrations-migrating-consuming-models

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

```console
FAIL: util_unix_test.go:105: UniterSuite.TestRunCommand
```

**Pull request merged**